### DISCUSSION TITLE
feature normalization

### DISCUSSION THREAD

**Discussion by mtzortzi** [pyg-team/pytorch_geometric#4388] [Q&A] [answered]
Hi all,

I'm using PyG to make graph regression in a problem. I would like to ask you the following question.
I have my targets which are two values per graph, and one is smaller than the other, for example, I have one value in the range of 0.4 and the other value which I have to predict together is in the range of 8000.
So to make it work I have to normalize them somehow. My question is, should I normalize them together per row? or should I normalize each one of them separately and after that combine them to make my training?

Thank you in advance

**Comment by rusty1s** [comment, marked as answer]
I think it is best to normalize each target separately, e.g., via
```python
mean, std = dataset.data.y.mean(dim=0, keepdim=True), dataset.data.y.std(dim=0, keepdim=True)

for batch in dataset:
    batch.y = (batch.y - mean) / std
```

**Reply by mtzortzi** [reply]
Great. 
So, by separately we mean, calculating for the whole dataset (before splitting into train, valid, test) the mean+std for the target y1 and normalizing it, and then we do the same thing for target y2?

**Reply by rusty1s** [reply]
Indeed. If `data.y` is of shape `[1, 2]`, then the above code will achieve this already.

**Reply by mtzortzi** [reply]
Thank you for your help :)

**Reply by Shiiivaa** [reply]
> Great. So, by separately we mean, calculating for the whole dataset (before splitting into train, valid, test) the mean+std for the target y1 and normalizing it, and then we do the same thing for target y2?

You need to do normalization after splitting data into train and test. Because test data are centered and scaled according to the training data.
